src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,332 @@
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
+            if (till < Ledger.CurrentIndex(engine.SnapshotCache) + 2) return false; // deposit must be valid at least until the next block after persisting block.
+            Deposit deposit = GetDepositFor(engine.SnapshotCache, addr);
+            if (deposit is null || till < deposit.Till) return false;
+            deposit.Till = till;
+
+            PutDepositFor(engine, addr, deposit);
+            return true;
+        }
+
+        /// <summary>
+        /// ExpirationOf returns deposit lock height for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit lock height of the specified address.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public uint ExpirationOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Till;
+        }
+
+        /// <summary>
+        /// BalanceOf returns deposited GAS amount for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit balance of the specified account.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public BigInteger BalanceOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Amount;
+        }
+
+        /// <summary>
+        /// Withdraw sends all deposited GAS for "from" address to "to" address. If "to"
+        /// address is not specified, then "from" will be used as a sender.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">From Account</param>
+        /// <param name="to">To Account</param>
+        /// <returns>Whether withdrawal was successfull.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private async ContractTask<bool> Withdraw(ApplicationEngine engine, UInt160 from, UInt160 to)
+        {
+            if (!engine.CheckWitnessInternal(from)) throw new InvalidOperationException(string.Format("Failed to check witness for {0}", from.ToString()));
+            var receive = to is null ? from : to;
+            var deposit = GetDepositFor(engine.SnapshotCache, from) ?? throw new InvalidOperationException(string.Format("Deposit of {0} is null", from.ToString()));
```

Comment:
Remove can return the deposit (interoperableClone) not related to the storage, and avoid get/remove

Reply:
> Remove can return the deposit 

But we're not interested in return value, it's just not needed, we're OK with the fact that deposit is removed.

> and avoid get/remove

We need to perform some validation checks before removal, otherwise it's illegal. Also, it's nice to have separate get/remove functions because they are reused from several Notary methods.